modules/nlu/src/backend/application/training-repo.ts

```diff
@@ -0,0 +1,138 @@
+import * as sdk from 'botpress/sdk'
+import { Transaction } from 'knex'
+import ms from 'ms'
+import { TrainingId, TrainingState, TrainingSession, I } from './typings'
+
+const TABLE_NAME = 'nlu_training_queue'
+const TRANSACTION_TIMEOUT_MS = ms('5s')
+const debug = DEBUG('nlu').sub('database')
+
+const TRANSACTION_TIMEOUT_ERROR = "Transaction exceeded it's time limit"
+
+const timeout = (ms: number) => {
+  return new Promise((_, reject) => {
+    setTimeout(() => reject(new Error(TRANSACTION_TIMEOUT_ERROR)), ms)
+  })
+}
+
+export type ITransactionContext = I<TransactionContext>
+class TransactionContext {
+  constructor(protected _database: typeof sdk.database, public transaction: Transaction | null = null) {}
+
+  private get table() {
+    if (this.transaction) {
+      return this._database.table(TABLE_NAME).transacting(this.transaction)
+    }
+    return this._database.table(TABLE_NAME)
+  }
+
+  public set = async (trainId: TrainingId, trainState: TrainingState): Promise<void> => {
+    const { botId, language } = trainId
+    const { progress, status, owner } = trainState
+
+    const modifiedOn = this._database.date.now()
+    if (await this.has({ botId, language })) {
+      return this.table.where({ botId, language }).update({ progress, status, modifiedOn, owner })
+    }
+    return this.table.insert({ botId, language, progress, status, modifiedOn, owner })
+  }
+
+  public has = async (trainId: TrainingId): Promise<boolean> => {
```

Comment:
what's the point of exposing a has method if it does the exact same thing as get. 

IMO `has` underlying query should return boolean and should not select anything from DB. 

Further more I just looked it up and the method in the repo that maps to this one isn't even used, making this dead code

Reply:
it's used by the set method to decide between an update and an insert (knex has no upsert methods that I know of))

Reply:
would you prefer a .count() instead of the whole query to check if it's in the database?